From 7e28ecba1a04c4956fb281802d6d27a60a0b68de Mon Sep 17 00:00:00 2001
Message-Id: <7e28ecba1a04c4956fb281802d6d27a60a0b68de.1435736732.git.chang-joon.lee@intel.com>
In-Reply-To: <7c8af138acbe907021f93af6d5e51d2a311ab097.1435736732.git.chang-joon.lee@intel.com>
References: <7c8af138acbe907021f93af6d5e51d2a311ab097.1435736732.git.chang-joon.lee@intel.com>
From: Gerard Andre <gerard.andre@intel.com>
Date: Thu, 25 Jun 2015 11:06:30 +0200
Subject: [PATCH 2/3] REVERTME [VPG]: drm/i915: Increase vblank evade timeout
 val for PUNIT write

This patch increases the evade timeout to 2msec after disabling maxfifo and
PM5 - which includes PUNIT writes. PUNIT needs atleast 2msec to serve
any request.

REVERTME: 2msec delay is required for PUNIT operations. However, in this
patch same delay is used for maxfifo writes as well. This needs further
optimization later.

Issue: IMINAN-33414
Change-Id: Ie7f2019e38ef7f30d418ccd2fd05f58919ea58e5
Signed-off-by: Arun R Murthy <arun.r.murthy@intel.com>
Signed-off-by: vandita kulkarni <vandita.kulkarni@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.h      | 2 ++
 drivers/gpu/drm/i915/intel_display.c | 4 +++-
 drivers/gpu/drm/i915/intel_pm.c      | 2 ++
 3 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 8b18c25..0d978f6 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -1811,6 +1811,8 @@ struct drm_i915_private {
 	/* Indicates currently enabled planes */
 	unsigned int pipe_plane_stat;
 	unsigned int prev_pipe_plane_stat;
+	/* delay in us */
+	unsigned int evade_delay;
 
 	/* PCH chipset type */
 	enum intel_pch pch_type;
diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index f7e8274..c72dcaa 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -11209,7 +11209,8 @@ static void intel_pipe_vblank_evade(struct drm_crtc *crtc)
 	struct intel_crtc *intel_crtc = to_intel_crtc(crtc);
 	int pipe = intel_crtc->pipe;
 	/* FIXME needs to be calibrated sensibly */
-	u32 min = crtc->mode.crtc_vdisplay - usecs_to_scanlines(crtc, 50);
+	u32 min = crtc->mode.crtc_vdisplay - usecs_to_scanlines(crtc,
+						dev_priv->evade_delay);
 	u32 max = crtc->mode.crtc_vdisplay - 1;
 	u32 val;
 
@@ -11230,6 +11231,7 @@ static void intel_pipe_vblank_evade(struct drm_crtc *crtc)
 		dev_warn(dev->dev,
 			 "Page flipping close to vblank start (DSL=%u, VBL=%u)\n",
 			 val, crtc->hwmode.crtc_vdisplay);
+	dev_priv->evade_delay = 2000;
 }
 
 static int intel_crtc_set_display(struct drm_crtc *crtc,
diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index 7abb7b4..61a5793 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -1699,6 +1699,7 @@ void intel_update_maxfifo(struct drm_i915_private *dev_priv,
 			vlv_punit_write(dev_priv, CHV_DPASSC,
 					(val | CHV_PW_MAXFIFO_MASK));
 			mutex_unlock(&dev_priv->rps.hw_lock);
+			dev_priv->evade_delay = 2000;
 		} else
 			I915_WRITE(FW_BLC_SELF_VLV, FW_CSPWRDWNEN);
 		dev_priv->maxfifo_enabled = true;
@@ -1721,6 +1722,7 @@ void intel_update_maxfifo(struct drm_i915_private *dev_priv,
 			I915_WRITE_BITS(DSPFW1, 0, 0xff800000);
 			I915_WRITE(DSPHOWM, (I915_READ(DSPHOWM) &
 							~(0x3000000)));
+			dev_priv->evade_delay = 2000;
 		} else
 			I915_WRITE(FW_BLC_SELF_VLV, ~FW_CSPWRDWNEN);
 		dev_priv->maxfifo_enabled = false;
-- 
1.9.1

